File: src/main/java/kitchenpos/product/tobe/ui/ProductRestController.java
package kitchenpos.product.tobe.ui;

import java.net.URI;
import java.util.List;
import java.util.UUID;
import javax.validation.Valid;
import kitchenpos.product.tobe.application.ProductService;
import kitchenpos.product.tobe.application.dto.ChangeProductPriceRequest;
import kitchenpos.product.tobe.application.dto.ChangeProductPriceResponse;
import kitchenpos.product.tobe.application.dto.CreateProductRequest;
import kitchenpos.product.tobe.application.dto.CreateProductResponse;
import kitchenpos.product.tobe.domain.Product;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RequestMapping("/api/products")
@RestController
public class ProductRestController {

    private final ProductService productService;

    public ProductRestController(final ProductService productService) {
        this.productService = productService;
    }

    @PostMapping
    public ResponseEntity<CreateProductResponse> create(@Valid @RequestBody final CreateProductRequest request) {
        final var response = productService.create(request);
        return ResponseEntity.created(URI.create("/api/products/" + response.getId()))
            .body(response);
    }

    @PutMapping("/{productId}/price")
    public ResponseEntity<ChangeProductPriceResponse> changePrice(@PathVariable final UUID productId,
        @Valid @RequestBody final ChangeProductPriceRequest request) {
        return ResponseEntity.ok(productService.changePrice(productId, request));
    }

    @GetMapping
    public ResponseEntity<List<Product>> findAll() {
        return ResponseEntity.ok(productService.findAll());
    }
}

File: src/main/kotlin/kitchenpos/product/tobe/ui/ProductRestController.kt
package kitchenpos.product.tobe.ui

import kitchenpos.product.tobe.application.ProductService
import kitchenpos.product.tobe.application.dto.ChangeProductPriceReq
import kitchenpos.product.tobe.application.dto.CreateProductReq
import kitchenpos.product.tobe.application.dto.ProductResp
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.net.URI
import java.util.*

@RequestMapping("/api/products")
@RestController
class ProductRestController(
    private val productService: ProductService
) {
    @PostMapping
    fun create(@RequestBody request: CreateProductReq): ResponseEntity<ProductResp> {
        val response = productService.create(request)
        return ResponseEntity.created(URI.create("/api/products/" + response.id))
            .body(response)
    }

    @PutMapping("/{productId}/price")
    fun changePrice(
        @PathVariable productId: UUID,
        @RequestBody request: ChangeProductPriceReq
    ): ResponseEntity<ProductResp> {
        return ResponseEntity.ok(productService.changePrice(productId, request))
    }

    @GetMapping
    fun findAll(): ResponseEntity<List<ProductResp>> {
        return ResponseEntity.ok(
            productService.findAll()
        )
    }
}
